src/Network.cpp

```diff
@@ -629,50 +629,46 @@ void Network::show_heatmap(FastState * state, Netresult& result, bool topmoves)
     }
 }
 
-void Network::gather_features(GameState * state, NNPlanes & planes) {
-    planes.resize(18);
-    constexpr size_t our_offset   = 0;
-    constexpr size_t their_offset = 8;
-    BoardPlane& black_to_move  = planes[16];
-    BoardPlane& white_to_move  = planes[17];
-
-    int to_move = state->get_to_move();
-    bool whites_move = to_move == FastBoard::WHITE;
-    if (whites_move) {
-        white_to_move.set();
-    } else {
-        black_to_move.set();
-    }
 
-    // Go back in time, fill history boards
-    size_t backtracks = 0;
-    for (int h = 0; h < 8; h++) {
-        // collect white, black occupation planes
-        for (int j = 0; j < 19; j++) {
-            for(int i = 0; i < 19; i++) {
-                int vtx = state->board.get_vertex(i, j);
-                FastBoard::square_t color =
-                    state->board.get_square(vtx);
-                int idx = j * 19 + i;
-                if (color != FastBoard::EMPTY) {
-                    if (color == to_move) {
-                        planes[our_offset + h][idx] = true;
-                    } else {
-                        planes[their_offset + h][idx] = true;
-                    }
+void Network::fill_input_plane(const FastBoard& board, InputPlane& plane) {
+    auto idx = 0;
+    for (int j = 0; j < 19; j++) {
+        for(int i = 0; i < 19; i++) {
+            int vtx = board.get_vertex(i, j);
+            auto color = board.get_square(vtx);
+            if (color != FastBoard::EMPTY) {
+                if (color == FastBoard::BLACK) {
+                    plane.first[idx] = true;
+                } else {
+                    plane.second[idx] = true;
                 }
             }
+            idx++;
         }
-        if (!state->undo_move()) {
-            break;
-        } else {
-            backtracks++;
-        }
+    }
+}
+
+void Network::gather_features(const GameState * state, NNPlanes & planes) {
+    planes.resize(INPUT_CHANNELS);
+    BoardPlane& black_to_move  = planes[16];
+    BoardPlane& white_to_move  = planes[17];
+
+    const auto blacks_move = state->board.black_to_move();
+
+    const auto black_offset = blacks_move ? 0 : INPUT_MOVES;
+    const auto white_offset = blacks_move ? INPUT_MOVES : 0;
+
+    if (blacks_move) {
+        black_to_move.set();
+    } else {
+        white_to_move.set();
     }
 
-    // Now go back to present day
-    for (size_t h = 0; h < backtracks; h++) {
-        state->forward_move();
+    const auto moves = std::min<int>(state->get_movenum(), INPUT_MOVES);
+    for (auto h = 0; h < moves; h++) {
+        const auto& test = state->get_boardplanes(h);
+        planes[black_offset + h] = test.first;
```

Comment:
this performs a copy of the bitset and is avoidable by changing NNPlane to use pointers to bitsets but that's a future change